backend/src/Types/Database.hs

```diff
@@ -1,117 +1,133 @@
-module Types.Database where
-
-import Data.Aeson (ToJSON)
-import Data.Time (UTCTime)
-import Database.SQLite.Simple (FromRow, ToRow)
-import Types.DataField
-  ( Competition,
-    EloId,
-    Expansion,
-    League,
-    Match,
-    PlayerId,
-    PlayerName,
-    Rating,
-    ReportId,
-    Side,
-    Stronghold,
-    Victory,
-  )
-
-data WriteProcessedGameReport = WriteProcessedGameReport
-  { timestamp :: UTCTime,
-    winnerId :: PlayerId,
-    loserId :: PlayerId,
-    side :: Side,
-    victory :: Victory,
-    match :: Match,
-    competition :: [Competition],
-    league :: Maybe League,
-    expansions :: [Expansion],
-    treebeard :: Maybe Bool,
-    actionTokens :: Int,
-    dwarvenRings :: Int,
-    turns :: Int,
-    corruption :: Int,
-    mordor :: Maybe Int,
-    initialEyes :: Int,
-    aragornTurn :: Maybe Int,
-    strongholds :: [Stronghold],
-    interestRating :: Int,
-    comments :: Maybe Text
-  }
-  deriving (Generic)
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
 
-instance ToRow WriteProcessedGameReport
-
-data ReadProcessedGameReport = ReadProcessedGameReport
-  { rid :: Int,
-    timestamp :: UTCTime,
-    winner :: PlayerName,
-    loser :: PlayerName,
-    side :: Side,
-    victory :: Victory,
-    match :: Match,
-    competition :: [Competition],
-    league :: Maybe League,
-    expansions :: [Expansion],
-    treebeard :: Maybe Bool,
-    actionTokens :: Int,
-    dwarvenRings :: Int,
-    turns :: Int,
-    corruption :: Int,
-    mordor :: Maybe Int,
-    initialEyes :: Int,
-    aragornTurn :: Maybe Int,
-    strongholds :: [Stronghold],
-    interestRating :: Int,
-    comments :: Maybe Text
-  }
-  deriving (Generic)
-
-instance FromRow ReadProcessedGameReport
+module Types.Database where
 
-instance ToJSON ReadProcessedGameReport
+import Data.Time (UTCTime (utctDay), Year, toGregorian)
+import Data.Time.Clock (getCurrentTime)
+import Database.Persist.TH (mkMigrate, mkPersist, persistLowerCase, share, sqlSettings)
+import Types.DataField (Competition, Expansion, League, Match, PlayerName, Rating, Side (..), Stronghold, Victory)
 
-data WritePlayer = WritePlayer
-  { name :: PlayerName,
-    country :: Maybe Text
-  }
-  deriving (Generic)
+share
```

Comment:
This definition of the database probably looks like nonsense - it's something called "Template Haskell" (TH) which is basically a DSL you use to write code that writes more Haskell for you. At compile time, this custom syntax is compiled into Haskell source code, and then *that* valid Haskell is compiled. A bit confusing, but great for auto-generating boilerplate that would otherwise have to all be hand-written.

This syntax comes from `persistent` which defines all the TH rules for defining a table. When this is compiled, it auto-generates a ton of data types for the records, their fields, and associated tables.

Reply:
This is serious magic 🤯

Reply:
YES